R/reactives.R

```diff
@@ -903,6 +909,125 @@ registerDebugHook("observerFunc", environment(), label)
+#' when \code{getCurrentObserver()} is called. If there is no observer being
+#' run (for example, if you called it from outside of a reactive context),
+#' it will always return \code{NULL}. There are a few subtleties, however.
+#' Consider the following five situations:
+#'
+#' \enumerate{
+#'   \item \code{getCurrentObserver()   #outside of a reactive context}
+#'   \item \code{observe({              getCurrentObserver()      }) }
+#'   \item \code{observe({ (function(){ getCurrentObserver() })() )} }
+#'   \item \code{observe({ isolate({    getCurrentObserver() })   }) }
+#'   \item \code{observe({ reactive({   getCurrentObserver() })   }) }
+#' }
+#'
+#' In (1), since you're outside of a reactive context, we've already
+#' established that \code{getCurrentObserver()} will return \code{NULL}.
+#' In (2), we have the "vanilla" case, in which \code{getCurrentObserver()}
+#' is called directly from within the body of the \code{observe} call.
+#' This returns that observer. So far, so good. The problem comes with
+#' the last three cases -- should we be able to "retrieve" the outer
+#' observer if we're inside an inner function's scope, or inside of an
+#' \code{isolate} or a \code{reactive} block?
+#'
+#' Before we can even asnwer that, there is an important distinction to
+#' be made here: are function calls, \code{reactive} calls and
+#' \code{isolate} blocks the same \emph{type} of thing? As far as Shiny
+#' is concerned, the answer is no. Shiny-specific things (like observers,
+#' reactives and code inside of an \code{isolate} chunk) exist in what we
+#' call reactive contexts. Each run of an observer or a reactive is
+#' associated with a particular reactive context. But regular functions
+#' have no relation to reactive contexts. So, while calling a regular
+#' function inside of an observer does not change the reactive context,
+#' calling a \code{reactive} or \code{isolate} certainly does.
+#'
+#' With this distinction in mind, we can refine our definition of
+#' \code{getCurrentObserver()} as follows: it returns the observer (if any)
+#' that is currently running, as long as it is called from within the
+#' same reactive context that was created when the observer started
+#' running. If the reactive context changed (most likely because of a
+#' call to \code{reactive} or \code{isolate}), \code{getCurrentObserver}
+#' will return \code{NULL}. (There is another common way that the reactive
+#' context can change inside an observer, which is if there is a second,
+#' nested observer. In this case, \code{getCurrentObserver()} will return
+#' the second, nested observer, since that is the one that is actually
+#' running at that time.)
+#'
+#' So to recap, here's the return value for each of the five situations:
+#' \enumerate{
+#'   \item \code{NULL}
+#'   \item the observer
+#'   \item the observer
+#'   \item \code{NULL}
+#'   \item \code{NULL}
+#' }
+#'
+#' Now, you may be wondering why \code{getCurrentObserver()} should't be able
```

Comment:
> It is also possible to look for the most recent observer in the reactive context "stack" by using `dig=TRUE`. The reactive context "stack" is analogous to a function call stack, except it tracks reactive contexts instead of function calls. With this option, when `getCurrentObserver()` is called from a `reactive` or `isolate`, it will return the observer at the bottom of the stack. (Because observers are never called by other observers, there can only be one observer in the stack.)